### PR TITLE
Added searchable encryption functionality

### PR DESCRIPTION
**Pull request by AmalieDue**
### Description
Added searchable encryption functionality.

Closes DEV-80

### Type(s) of Change (Split the PR if you check many)

- [ ] Added user feature
- [x] Added technical feature
- [x] Added tests
- [ ] Refactor
- [ ] Bugfix
- [ ] Updated documentation
- [ ] Kubernetes changes
- [ ] CI/CD workflow changes

### Testing Checklist

- [x] I have added/updated unit tests for functional changes.
- [ ] I have added/updated integration tests for changes that affect dependent systems.
- [ ] I have added/updated end-to-end tests for feature changes.

### General Checklist

- [x] I have pulled in the latest changes from master.
- [x] I have run `make lint`.
- [x] I have successfully run `make tests`.
- [ ] I have updated relevant CI/CD workflows.
- [ ] I have updated relevant Kubernetes manifests/scripts.
- [ ] I have updated/added relevant (internal and external) documentation (readme, doc comments, etc).
- [ ] I have updated the dependency map to reflect my changes.
- [ ] I have added the license to new source code files.
- [x] I have spent some time looking over the full diff before creating this PR.

### CODE REVIEW

File: searchable.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+// SealedID is an encrypted structure which defines an occurrence of a specific keyword in a specific ID.
+type SealedID struct {
+	ciphertext []byte
+	wrappedKey []byte
+}
+
+const MasterKeyLength = 32
+
+var ErrInvalidMasterKeyLength = fmt.Errorf("invalid key length, accepted key length is %d bytes", MasterKeyLength)
```
Comment (philip-cc) [inline searchable.go:38]: It's convention to keep package global constants/variables near the top of the file (after the imports).

File: searchable.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+// NewSearchable creates an Index which is used to manage keyword/ID pairs.
+func NewSearchable() Index {
```
Comment (philip-cc) [inline searchable.go:41]: To follow convention this should be called `NewIndex`.

File: searchable.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+	mapping := make(map[string]SealedID)
+
+	return Index{mapping: mapping}
```
Comment (philip-cc) [inline searchable.go:44]:
```suggestion
	return Index{mapping: make(map[string]SealedID)
}
```

File: searchable.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+// Add is used to add a keyword/ID pair to the Index.
+func (i *Index) Add(key []byte, keyword, id string) error {
+	if len(key) != MasterKeyLength {
+		return ErrInvalidMasterKeyLength
+	}
+
+	k1 := crypto.KMACKDF(crypto.TaggerKeyLength, key, []byte("mac"), []byte(keyword))
```
Comment (philip-cc) [inline searchable.go:53]:
```suggestion
	k1 := crypto.KMACKDF(crypto.TaggerKeyLength, key, []byte("label"), []byte(keyword))
```

File: searchable.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+// NewSearchable creates an Index which is used to manage keyword/ID pairs.
+func NewSearchable() Index {
+	mapping := make(map[string]SealedID)
+
+	return Index{mapping: mapping}
+}
+
+// Add is used to add a keyword/ID pair to the Index.
+func (i *Index) Add(key []byte, keyword, id string) error {
```
Comment (philip-cc) [inline searchable.go:48]: Some comments in this function to explain what the purpose of the different steps are would be great.

File: searchable.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+		if encryptedID, ok := i.mapping[base64.StdEncoding.EncodeToString(label)]; !ok {
+			break
+		} else {
+			var plaintext []byte
+			err := cryptor.Decrypt(&plaintext, "", encryptedID.wrappedKey, encryptedID.ciphertext)
+			if err != nil {
+				return nil, err
+			}
+			decryptedIDs = append(decryptedIDs, string(plaintext))
+		}
```
Comment (philip-cc) [inline searchable.go:120]:
```suggestion
		encryptedID, ok := i.mapping[base64.StdEncoding.EncodeToString(label)]
		if !ok {
			break
		}

		var plaintext []byte
		err := cryptor.Decrypt(&plaintext, "", encryptedID.wrappedKey, encryptedID.ciphertext)
		if err != nil {
			return nil, err
		}
		decryptedIDs = append(decryptedIDs, string(plaintext))
```

File: searchable.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+	count, err := i.count(key, keyword)
+	if err != nil {
+		return err
+	}
+	label, err := tagger.Tag(uint64Encode(count))
+	if err != nil {
+		return err
+	}
+
+	wrappedKey, encryptedID, err := cryptor.Encrypt([]byte(id), "")
```
Comment (philip-cc) [inline searchable.go:74]: You should just be able to do `cryptor.Encrypt(&id, "")` and then similarly pass a `*string` to `Decrypt`.

File: searchable.go [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+	for j := 0; ; j++ {
+		label, err := tagger.Tag(uint64Encode(uint64(j)))
+		if err != nil {
+			return 0, err
+		}
+
+		if _, ok := i.mapping[base64.StdEncoding.EncodeToString(label)]; !ok {
+			count = uint64(j)
+			break
```
Comment (philip-cc) [inline searchable.go:148]: You can simplify this a bit (and maybe remove the return at the end, if the compiler accepts it). 
```suggestion
			return uint64(j)
```

File: searchable_test.go [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+}
+
+func TestAdd(t *testing.T) {
+	searchable := NewSearchable()
+
+	rand := &crypto.NativeRandom{}
+	masterKey, err := rand.GetBytes(32)
+	if err != nil {
+		t.Fatalf("Random failed: %v", err)
+	}
```
Comment (philip-cc) [inline searchable_test.go:37]: It's up to you, but I'd say it's okay to skip the error handling (in tests!) for things you "know" cannot fail. 
```suggestion
	masterKey, _ := rand.GetBytes(32)
```

File: searchable_test.go [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+
+	keyword := "first keyword"
+	id := "first id"
+
+	err = searchable.Add(masterKey, keyword, id)
+	if err != nil {
```
Comment (philip-cc) [inline searchable_test.go:43]: I like using the compact syntax in these cases (here and elsewhere):
```suggestion
	if err := searchable.Add(masterKey, keyword, id); err != nil {
```

File: searchable.go (hunk 1 above)
Comment (AmalieDue) [reply searchable.go]: Fixed

File: searchable.go (hunk 2 above)
Comment (AmalieDue) [reply searchable.go]: Fixed

File: searchable.go (hunk 5 above)
Comment (AmalieDue) [reply searchable.go:46]: Fixed

File: searchable.go (hunk 7 above)
Comment (AmalieDue) [reply searchable.go]: Fixed

**Review by philip-cc — APPROVED** [review]

**Pull request merged**